### 异步的概念
异步（Asynchronous, async）是与同步（Synchronous, sync）相对的概念。
简单来理解就是：同步按你的代码顺序执行，异步不按照代码顺序执行，异步的执行效率更高。
通俗地解释一下异步：异步就是从主线程发射一个子线程来完成任务。

### 什么时候用异步编程
当浏览器在处理一个页面时，如果主线程遇到一个耗时的操作，比如读取一个文件或者发出一个网络请求，那么这个操作就会耗费时间，如果这个操作没有结束，那么浏览器就无法处理其他请求。

为了避免这种情况的发生，常常用子线程来完成耗时操作。因为子线程独立于主线程，所以即使出现阻塞也不会影响主线程的运行。

但是子线程有一个局限：一旦发射了以后就会与主线程失去同步，无法确定它的结束，如果结束之后需要处理一些事情，比如处理来自服务器的信息，我们是无法将它合并到主线程中去的。

为了解决这个问题，JavaScript 中的异步操作函数往往通过回调函数来实现异步任务的结果处理。

### 回调函数
回调函数就是一个函数，它是在我们启动一个异步任务的时候就告诉它：等你完成了这个任务之后要干什么。这样一来主线程几乎不用关心异步任务的状态了，它自己会善始善终。

```js
// 实例
function print() {
    document.getElementById("demo").innerHTML="RUNOOB!";
}

// 异步函数
setTimeout(print, 3000);
```

### Promise
Promise是一个ECMAScript6提供的类，目的是更加优雅地书写复杂的异步任务。

#### 构造Promise
新建一个Promise对象：

```js
new Promise(function (resolve, reject) {
    // 要做的事情...
});
```

通过新建一个Promise对象好像并没有看出它怎样 "更加优雅地书写复杂的异步任务"。我们之前遇到的异步任务都是一次异步，如果需要多次调用异步函数呢？例如，如果我想分三次输出字符串，第一次间隔 1 秒，第二次间隔 4 秒，第三次间隔 3 秒：

```js
// 实例
setTimeout(function () {
    console.log("First");
    setTimeout(function () {
        console.log("Second");
        setTimeout(function () {
            console.log("Third");
        }, 3000);
    }, 4000);
}, 1000);
```

这段程序实现了这个功能，但是它是用 "函数瀑布" 来实现的。可想而知，在一个复杂的程序当中，用 "函数瀑布" 实现的程序无论是维护还是异常处理都是一件特别繁琐的事情，而且会让缩进格式变得非常冗赘。

现在我们用Promise来实现同样的功能：

```js
// 实例
new Promise(function (resolve, reject) {
    setTimeout(function () {
        console.log("First");
        resolve();
    }, 1000);
}).then(function () {
    return new Promise(function (resolve, reject) {
        setTimeout(function () {
            console.log("Second");
            resolve();
        }, 4000);
    });
}).then(function () {
    setTimeout(function () {
        console.log("Third");
    }, 3000);
});
```

这段代码较长，所以还不需要完全理解它，我想引起注意的是Promise将嵌套格式的代码变成了顺序格式的代码。

#### Promise的构造函数
Promise构造函数是JavaScript中用于创建Promise对象的内置构造函数。

Promise构造函数接受一个函数作为参数，该函数是同步的并且会被立即执行，所以我们称之为起始函数。起始函数包含两个参数```resolve```和```reject```，分别表示Promise```成功和失败```的状态。

起始函数执行成功时，它应该调用```resolve```函数并传递成功的结果。当起始函数执行失败时，它应该调用```reject```函数并传递失败的原因。

Promise构造函数返回一个Promise对象，该对象具有以下几个方法：
- then：用于处理Promise成功状态的回调函数。
- catch：用于处理Promise失败状态的回调函数。
- finally：无论Promise是成功还是失败，都会执行的回调函数。

下面是一个使用Promise构造函数创建Promise对象的例子：
当Promise被构造时，起始函数会被同步执行：

```js
// 实例
const promise = new Promise((resolve, reject) => {
  // 异步操作
  setTimeout(() => {
    if (Math.random() < 0.5) {
      resolve('success');
    } else {
      reject('error');
    }
  }, 1000);
});
 
promise.then(result => {
  console.log(result);
}).catch(error => {
  console.log(error);
});
```

在上面的例子中，我们使用Promise构造函数创建了一个Promise对象，并使用setTimeout模拟了一个异步操作。如果异步操作成功，则调用```resolve```函数并传递成功的结果；如果异步操作失败，则调用```reject```函数并传递失败的原因。然后我们使用```then```方法处理Promise成功状态的回调函数，使用```catch```方法处理Promise失败状态的回调函数。

但是请注意以下两点：
- resolve和reject 的作用域只有起始函数，不包括then以及其他序列；
- resolve和reject 并不能够使起始函数停止运行，别忘了return。

#### Promise函数
上述的 "计时器" 程序看上去比函数瀑布还要长，所以我们可以将它的核心部分写成一个```Promise函数```：

```js
// 实例
function print(delay, message) {
    return new Promise(function (resolve, reject) {
        setTimeout(function () {
            console.log(message);
            resolve();
        }, delay);
    });
}
```

然后我们就可以放心大胆的实现程序功能了：

```js
// 实例
print(1000, "First").then(function () {
    return print(4000, "Second");
}).then(function () {
    print(3000, "Third");
});
```

这种返回值为一个Promise对象的函数称作```Promise函数```，它常常用于开发基于异步操作的库。

#### 异步函数
异步函数（async function）是ECMAScript 2017 (ECMA-262) 标准的规范，几乎被所有浏览器所支持，除了Internet Explorer。

在Promise中我们编写过一个Promise函数：

```js
// 实例
function print(delay, message) {
    return new Promise(function (resolve, reject) {
        setTimeout(function () {
            console.log(message);
            resolve();
        }, delay);
    });
}
```

然后用不同的时间间隔输出了三行文本：

```js
// 实例
print(1000, "First").then(function () {
    return print(4000, "Second");
}).then(function () {
    print(3000, "Third");
});
```

可以将这段代码变得更好看：

```js
// 实例
async function asyncFunc() {
    await print(1000, "First");
    await print(4000, "Second");
    await print(3000, "Third");
}
asyncFunc();
```

这样将异步操作变得像同步操作一样容易了！
这次的回答是肯定的，异步函数（async function）中可以使用```await```指令，```await```指令后必须跟着一个Promise，异步函数会在这个Promise运行中暂停，直到其运行结束再继续运行。

处理异常的机制将用try-catch块实现：

```js
实例
async function asyncFunc() {
    try {
        await new Promise(function (resolve, reject) {
            throw "Some error"; // 或者 reject("Some error")
        });
    } catch (err) {
        console.log(err);
        // 会输出 Some error
    }
}
asyncFunc();
```

如果Promise有一个正常的返回值，await语句也会返回它：

```js
// 实例
async function asyncFunc() {
    let value = await new Promise(
        function (resolve, reject) {
            resolve("Return value");
        }
    );
    console.log(value);
}
asyncFunc();
// 程序会输出: Return value
```